SystemAudioPlayer/impl/AudioPlayer.cpp

```diff
@@ -90,27 +93,49 @@ void AudioPlayer::Init(SAPEventCallback *callback)
     SAPLOG_INFO("SAP: AudioPlayer Init\n");
     if(!gst_is_initialized())
         gst_init(NULL,NULL);
+
     m_main_loop_thread = g_thread_new("BusWatch", (void* (*)(void*)) event_loop, NULL);
+    waitForLoopStart();
     m_callback = callback;
     Soc_Initialize();
 }
 
-void AudioPlayer::DeInit()
-{   
-    SAPLOG_INFO("SAP: AudioPlayer DeInit\n");
-    if(g_main_loop_is_running(m_main_loop))
-        g_main_loop_quit(m_main_loop);
-    g_thread_join(m_main_loop_thread);
-    Soc_Deinitialize();
-    SAPLOG_INFO("SAP: AudioPlayer DeInit last\n");
+void AudioPlayer::waitForLoopStart()
+{
+    std::unique_lock<std::mutex> lock(m_eventMutex);
+    while (!m_isLoopStarted) {
+        m_eventCondition.wait(lock);
+    }
 }
 
 void AudioPlayer::event_loop()
 {
     m_main_loop = g_main_loop_new(NULL, false);
+    g_idle_add(reinterpret_cast<GSourceFunc>(notifyLoopStarted),NULL);
     g_main_loop_run(m_main_loop);
 }
 
+gboolean AudioPlayer::notifyLoopStarted()
+{
+    std::unique_lock<std::mutex> lock(m_eventMutex);
+    m_isLoopStarted = true;
+    m_eventCondition.notify_one();
+    return G_SOURCE_REMOVE;
+}
+
+void AudioPlayer::DeInit()
+{
+    SAPLOG_INFO("SAP: AudioPlayer DeInit\n");
+    if(m_isLoopStarted && g_main_loop_is_running(m_main_loop))
```

Comment:
I think it is better to check for the validity of m_main_loop & if m_main_loop is running. Because with the current logic, it is possible for the Deinit() to be called after g_main_loop_run() is called but before `m_isLoopStarted` is set from the idle source.

Reply:
done